### PR TITLE
2020 schedule

### LINKED ISSUE #160: Prepare for sweeping schdule changes
MICDS has made drastic changes to the flow of a school day, so we're gonna need to adjust our systems

### PR DESCRIPTION
**Pull request by MeMeBigProgrammer**
resolves #160

### CODE REVIEW

File: src/libs/schedule.ts [hunk 1]
```diff
@@ -151,20 +133,33 @@ const genericBlocks: Record<
 	}
 };
 
+function isOnLateStartList(date: Date) {
+	const lateStartList = lateStarts as DateList;
+	lateStartList.date.forEach(value => {
```
Comment (nickbclifford) [inline src/libs/schedule.ts:138]: Can you turn this into a `for`..`of` loop?

File: src/libs/portal.ts [hunk 2]
```diff
@@ -336,48 +378,26 @@ export async function getFromCalCalendar(db: Db, user: string) {
...
-		// Check if it's an all-day event
-		if (startTime === scheduleDate.getTime() && Number.isNaN(endTime)) {
-			// See if valid day
-			if (validDayRotationPlain.test(calEvent.summary)) {
-				// Get actual day
-				return parseInt(/Day ([1-6])/.exec(calEvent.summary)![1], 10);
-			}
-		}
-	}
+	return withCalSummary(date, (summary: string) => {
```
Comment (nickbclifford) [inline src/libs/portal.ts:384]: Can you remove the type annotation? Also look at making this just a single-line arrow function, but feel free to keep it as bracketed if it's too ugly

File: src/libs/portal.ts [hunk 3]
```diff
@@ -336,48 +378,26 @@ export async function getFromCalCalendar(db: Db, user: string) {
...
 
-	return null;
+/**
+ * Return whether date is a late start
+ * @param Date the day to check
+ * @returns boolean of whether it is a late start day
+ */
+export async function isLateStart(date: Date) {
+	return (
+		(await withCalSummary(date, (summary: string) => {
```
Comment (nickbclifford) [inline src/libs/portal.ts:396]: ditto as above, also please extract this function call into its own variable just to clean up the return statement a bit

File: test/schedule.test.ts [hunk 4]
```diff
@@ -158,12 +156,12 @@ describe('Schedule', () => {
...
-					special: false,
+					day: 'E',
+					special: true, // COVID TODO things are broken rn, just go with it
 					classes: [
-						{ class: { name: 'alias class' } },
-						{ class: { name: 'Test Class 5' } },
-						{ class: { name: 'Test Class 3' } }
+						{ class: { name: 'Test Class 4' } },
+						{ class: { name: 'Test Class 2' } },
+						{ class: { name: 'alias class' } }
```
Comment (nickbclifford) [inline test/schedule.test.ts:164]: This test should be returning the same classes in the same order as the previous test, just with aliases included.

File: src/libs/schedule.ts (hunk 1 above)
Comment (MeMeBigProgrammer) [reply src/libs/schedule.ts:132]: aight

File: src/libs/portal.ts [hunk 5]
```diff
@@ -336,48 +378,28 @@ export async function getFromCalCalendar(db: Db, user: string) {
...
+ */
+export async function isLateStart(date: Date) {
+	// [A - G] because there 7 late start schedules (WHY???)
 
-		const startTime = start.getTime();
-		const endTime = end.getTime();
+	const lateStart =
+		(await withCalSummary(date, summary => {
+			return (/[A-G]9 Day/.exec(summary) ?? []).length > 0;
+		})) ?? false;
```
Comment (nickbclifford) [inline src/libs/portal.ts:400]: Sorry, I wasn't clear—I meant moving the `await withCalSummary()` call into its own variable so we don't need parentheses around it in the return statement.

File: src/libs/schedule.ts [hunk 6]
```diff
@@ -151,106 +127,151 @@ const genericBlocks: Record<
 	}
 };
 
+function isOnLateStartList(date: Date) {
```
Comment (nickbclifford) [inline src/libs/schedule.ts:130]: TypeDoc comment here, please

File: src/libs/schedule.ts [hunk 7]
```diff
@@ -151,106 +127,151 @@ const genericBlocks: Record<
...
+	for (const lateStartDate of lateStartList.date) {
+		const lateStartMoment = moment(lateStartDate, 'MM/DD/YYYY');
+		if (moment(date).isSame(lateStartMoment)) {
+			return true;
+		}
+	}
+	return false;
+}
+
+function removeCertainLunchBlocks(
```
Comment (nickbclifford) [inline src/libs/schedule.ts:141]: ditto

File: src/libs/schedule.ts [hunk 8]
```diff
@@ -151,106 +127,151 @@ const genericBlocks: Record<
...
+	daySchedule: blockSchedule.BlockFormats,
+	keepAemsh: boolean,
+	keepHswl: boolean,
+	keepDefault: boolean
```
Comment (nickbclifford) [inline src/libs/schedule.ts:145]: Consider refactoring this into a parameters object `{ keepAemsh, keepHswl, keepDefault }`, as it's a bit confusing when called.

File: src/libs/schedule.ts [hunk 9]
```diff
@@ -412,138 +423,72 @@ async function getSchedule(
...
+	return { hasURL: true, schedule: userSchedule };
+}
 
-			schedule.classes[i] = {
-				class: (scheduleClass as any).class as ScheduleClass,
-				start: scheduleClass.start as moment.Moment,
-				end: scheduleClass.end as moment.Moment
-			};
-		}
+function convertTimeStringToMoment(
```
Comment (nickbclifford) [inline src/libs/schedule.ts:481]: ditto

File: test/schedule.test.ts [hunk 10]
```diff
@@ -97,14 +96,13 @@ describe('Schedule', () => {
...
+					day: 'E',
 					special: false,
 					classes: [
-						{ class: { name: 'b block class' } },
-						{ class: { name: 'f block class' } },
-						{ class: { name: 'Collaborative Work' } },
-						{ class: { name: 'Activities' } },
-						{ class: { name: 'Block D' } }
+						{ class: { name: 'g block class' } },
+						{ class: { name: 'e block class' } },
```
Comment (nickbclifford) [inline test/schedule.test.ts:103]: Are these in the correct order? It should be the same as the previous test.

File: src/libs/schedule.ts (hunk 6 above)
Comment (MeMeBigProgrammer) [reply src/libs/schedule.ts:130]: no

File: test/schedule.test.ts (hunk 10 above)
Comment (MeMeBigProgrammer) [reply test/schedule.test.ts:103]: good catch

**Comment by MeMeBigProgrammer** [comment]
please don't break anything... please don't break anything...

**Pull request merged**